pandas/_libs/byteswap.pyx

```diff
@@ -10,53 +10,59 @@ from libc.stdint cimport (
     uint32_t,
     uint64_t,
 )
+from libc.string cimport memcpy
+
+
+cdef union Cast32:
+    uint32_t u
+    float f
+
+
+cdef union Cast64:
+    uint64_t u
+    double d
 
 
 def read_float_with_byteswap(bytes data, Py_ssize_t offset, bint byteswap):
-    assert offset + 4 < len(data)
-    cdef:
-        const char *data_ptr = data
-        float res = (<float*>(data_ptr + offset))[0]
+    cdef Cast32 res
+    assert offset + sizeof(res) < len(data)
+    memcpy(&res.u, <const unsigned char*>(data) + offset, sizeof(res))
```

Comment:
Can you not just memcpy to a float pointer and return that? Still unclear why we need the union